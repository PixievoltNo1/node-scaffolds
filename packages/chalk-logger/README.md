<header>
  <h1 align="center">
    <a href="https://github.com/guanghechen/node-scaffolds/tree/release-4.x.x/packages/chalk-logger#readme">@guanghechen/chalk-logger</a>
  </h1>
  <div align="center">
    <a href="https://www.npmjs.com/package/@guanghechen/chalk-logger">
      <img
        alt="Npm Version"
        src="https://img.shields.io/npm/v/@guanghechen/chalk-logger.svg"
      />
    </a>
    <a href="https://www.npmjs.com/package/@guanghechen/chalk-logger">
      <img
        alt="Npm Download"
        src="https://img.shields.io/npm/dm/@guanghechen/chalk-logger.svg"
      />
    </a>
    <a href="https://www.npmjs.com/package/@guanghechen/chalk-logger">
      <img
        alt="Npm License"
        src="https://img.shields.io/npm/l/@guanghechen/chalk-logger.svg"
      />
    </a>
    <a href="#install">
      <img
        alt="Module Formats: cjs"
        src="https://img.shields.io/badge/module_formats-cjs-green.svg"
      />
    </a>
    <a href="https://github.com/nodejs/node">
      <img
        alt="Node.js Version"
        src="https://img.shields.io/node/v/@guanghechen/chalk-logger"
      />
    </a>
    <a href="https://github.com/facebook/jest">
      <img
        alt="Tested with Jest"
        src="https://img.shields.io/badge/tested_with-jest-9c465e.svg"
      />
    </a>
    <a href="https://github.com/prettier/prettier">
      <img
        alt="Code Style: prettier"
        src="https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square"
      />
    </a>
  </div>
</header>
<br/>


`chalk-logger` is a colorful logger tool based on [chalk][] (so you can use a
lot of colors), and can be easily integrated into [commander.js][] (so you can
use command line parameters to customized the logger's behavior).


## Install

* npm

  ```bash
  npm install --save @guanghechen/chalk-logger
  ```

* yarn

  ```bash
  yarn add @guanghechen/chalk-logger
  ```

## Usage

### Options
Name                | Type                    | Required  | Default           |  Desc
:------------------:|:-----------------------:|:---------:|:-----------------:|:---------------------------------------------:
`basename`          | string \| `null`        | `false`   | `null`            | [see below](#option-details)
`mode`              | `'normal'` \| `'loose'` | `false`   | `normal`          | [see below](#option-details)
`placeholderRegex`  | RegExp                  | `false`   | `/(?<!\\)\{\}/g`  | string formatter placeholder regex
`name`              | string                  | `false`   | -                 | name of logger
`level`             | Level                   | `false`   | `Level.INFO`      | verbosity level of the logging output
`flags`             | See below               | `false`   | -                 | feature flags
`write`             | (text: string) => void  | `false`   | `process.stdout`  | [see below](#option-details)

### Option Details

* `basename`: Base of the logger name, when you change logger name according
  `setName` function, the basename will be prefixed of the logger name.

* `mode`

  - `normal`: Print log only
  - `loose`: Print a newline before and after the log

* `flags`

  ```typescript
  interface ILoggerFlags {
    date?: boolean
    title?: boolean
    inline?: boolean
    colorful?: boolean
  }
  ```

  Flag        | Type    | Required  | Default |  Desc
  :----------:|:-------:|:---------:|:-------:|:---------------------------------------------:
  `date`      | boolean | `false`   | `false` | whether to print the date
  `title`     | boolean | `false`   | `true`  | whether to print the title
  `inline`    | boolean | `false`   | `false` | whether to print each log on one line
  `colorful`  | boolean | `false`   | `true`  | whether to print log surrounded with color

* `write`: If `filepath` is specified, the log is output to `filepath` by default,
  otherwise to the `process.stdout`.  You can specify your own write function to
  customize the output behavior of the log.


### Cli Options

* `--log-level <debug|verbose|info|warn|error|fatal>`: specify global logger level.

* `--log-name <new logger name>`: specify global logger name.

* `--log-mode <'normal' | 'loose'>`: specify global logger mode.

* `--log-flag <[no-](date|title|inline|colorful)>`: the prefix `no-` represent negation.

  - `date`: whether to print date. default value is false
  - `title`: whether to print title. default value is true
  - `inline`: each log record output in one line. default value is false.
  - `colorful`: whether to print with colors. default value is true.


## Test

Use [@guanghechen/helper-jest][] to spy logger.

```typescript
import { ChalkLogger, Level } from '@guanghechen/chalk-logger'
import {
  composeStringDesensitizers,
  createFilepathDesensitizer,
  createJsonDesensitizer,
  createLoggerMock,
} from '@guanghechen/helper-jest'

const workspaceRootDir = path.resolve(__dirname, '..')
const desensitize = createJsonDesensitizer({
  string: composeStringDesensitizers(
    createFilepathDesensitizer(workspaceRootDir, '<$WORKSPACE$>'),
    text => text.replace(/\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}/, '<$Date$>'),
  ),
})

const logger = new ChalkLogger({ name: 'demo', level: Level.DEBUG, flags: { date: true } })
const loggerMock = createLoggerMock({ logger, desensitize })

// collect data
loggerMock.getIndiscriminateAll()

// reset mock
loggerMock.reset()

// restore mock
loggerMock.restore()
```


## Examples

* Basic:

  ```typescript
  // demo/demo1.ts
  import { ChalkLogger, Level } from '@guanghechen/chalk-logger'

  const logger = new ChalkLogger(
    {
      name: 'demo1',
      level: Level.ERROR, // the default value is INFO
      flags: {
        date: false, // the default value is false.
        colorful: true, // the default value is true.
      },
    },
    process.argv,
  )

  logger.debug('A', 'B', 'C')
  logger.verbose('A', 'B', 'C')
  logger.info('a', 'b', 'c')
  logger.warn('X', 'Y', 'Z', { a: 1, b: 2 })
  logger.error('x', 'y', 'z', { c: { a: 'hello' }, b: { d: 'world' } })
  logger.fatal('1', '2', '3')
  ```

  ![demo1.1.png][]

* Custom output format:

  ```typescript
  // demo/demo2.ts
  import { ChalkLogger, Level } from '@guanghechen/chalk-logger'

  const logger = new ChalkLogger(
    {
      name: 'demo2',
      level: Level.ERROR, // the default value is INFO
      flags: {
        date: false, // the default value is false.
        colorful: true, // the default value is true.
      },
    },
    process.argv,
  )

  logger.formatHeader = 
  function formatHeader(level: Level, date: Date): string {
      const dateText: string = this.flags.date
        ? this.formatContent(level, date.toLocaleTimeString())
        : ''

      const levelStyle = this.levelStyleMap[level]
      let levelText = levelStyle.title
      if (this.flags.colorful) {
        levelText = levelStyle.labelChalk.fg(levelText)
        if (levelStyle.labelChalk.bg != null) levelText = levelStyle.labelChalk.bg(levelText)
      }

      const titleText: string = this.flags.title
        ? this.formatContent(level, '<' + this.name + '>')
        : ''

      let result = ''
      if (dateText) result += dateText + ' '
      result += levelText
      if (titleText) result += ' ' + titleText
      return result
    }

  logger.debug('A', 'B', 'C')
  logger.verbose('A', 'B', 'C')
  logger.info('a', 'b', 'c')
  logger.warn('X', 'Y', 'Z', { a: 1, b: 2 })
  logger.error('x', 'y', 'z', { c: { a: 'hello' }, b: { d: 'world' } })
  logger.fatal('1', '2', '3')
  ```

  ![demo2.1.png][]


* Custom colors with [chalk][]:

  ```typescript
  // demo/demo3.ts
  import { ChalkLogger, Level } from '@guanghechen/chalk-logger'

  const logger = new ChalkLogger(
    {
      name: 'demo3',
      level: Level.ERROR, // the default value is INFO
      flags: {
        date: false, // the default value is false.
        colorful: true, // the default value is true.
      },
    },
    process.argv,
  )

  logger.debug('A', 'B', 'C')
  logger.verbose('A', 'B', 'C')
  logger.info('a', 'b', 'c')
  logger.warn('X', 'Y', 'Z', { a: 1, b: 2 })
  logger.error('x', 'y', 'z', { c: { a: 'hello' }, b: { d: 'world' } })
  logger.fatal('1', '2', '3')
  ```

  ![demo3.1.png][]

* Output to file

  ```typescript
  // demo/demo4.ts
  import { ChalkLogger, Level } from '@guanghechen/chalk-logger'
  import fs from 'node:fs'
  import path from 'node:path'
  import url from 'node:url'

  const __dirname = path.dirname(url.fileURLToPath(import.meta.url))

  const logFilepath = path.resolve(__dirname, 'orz.log')
  const logger = new ChalkLogger(
    {
      name: 'demo4',
      level: Level.DEBUG, // the default value is DEBUG
      flags: {
        date: true, // the default value is false.
        inline: true,
        colorful: false, // the default value is true.
      },
      write: text => fs.appendFileSync(logFilepath, text, 'utf-8'),
    },
    process.argv,
  )

  logger.debug('A', 'B', 'C')
  logger.verbose('A', 'B', 'C')
  logger.info('a', 'b', 'c')
  logger.warn('X', 'Y', 'Z', { a: 1, b: 2 })
  logger.error('x', 'y', 'z', { c: { a: 'hello' }, b: { d: 'world' } })
  logger.fatal('1', '2', '3')
  ```

  ![demo4.1.png][]

* With [commander.js][]:

  ```typescript
  // demo/demo5.ts
  import { ChalkLogger, Level, registerCommanderOptions } from '@guanghechen/chalk-logger'
  import { Command } from 'commander'

  const logger = new ChalkLogger(
    {
      name: 'demo5',
      level: Level.ERROR, // the default value is INFO
      flags: {
        date: false, // the default value is false.
        colorful: true, // the default value is true.
      },
    },
    process.argv,
  )

  const command = new Command()
  command.version('v1.0.0').arguments('[orz]')

  // register logger option to commander
  registerCommanderOptions(command)

  command.option('-e, --encoding <encoding>', "specified <filepath>'s encoding").parse(process.argv)

  logger.debug('A', 'B', 'C')
  logger.verbose('A', 'B', 'C')
  logger.info('a', 'b', 'c')
  logger.warn('X', 'Y', 'Z', { a: 1, b: 2 })
  logger.error('x', 'y', 'z', { c: { a: 'hello' }, b: { d: 'world' } })
  logger.fatal('1', '2', '3')
  ```

  ![demo5.1.png][]

* String format:

  ```typescript
  // demo/demo6.ts
  import { ChalkLogger, Level } from '@guanghechen/chalk-logger'

  const logger = new ChalkLogger(
    {
      name: 'demo6',
      level: Level.DEBUG,
      flags: {
        date: true,
        colorful: true,
        inline: true,
      },
    },
    process.argv,
  )

  logger.verbose('user({})', {
    username: 'lemon-clown',
    avatar:
      'https://avatars0.githubusercontent.com/u/42513619?s=400&u=d878f4532bb5749979e18f3696b8985b90e9f78b&v=4',
  })
  logger.error('bad argument ({}). error({})', { username: 123 }, new Error('username is invalid'))

  const logger2 = new ChalkLogger(
    {
      name: 'demo6',
      level: Level.DEBUG,
      flags: {
        date: true,
        colorful: true,
        inline: true,
      },
      placeholderRegex: /(?<!\\)<>/g, // change placeholder of string format
    },
    process.argv,
  )

  logger2.verbose('user(<>)', {
    username: 'lemon-clown',
    avatar:
      'https://avatars0.githubusercontent.com/u/42513619?s=400&u=d878f4532bb5749979e18f3696b8985b90e9f78b&v=4',
  })
  logger2.error('bad argument (<>). error({})', { username: 123 }, new Error('username is invalid'))
  ```

  ![demo6.1.png][]

* No title:

  ```typescript
  import { ChalkLogger, Level } from '@guanghechen/chalk-logger'

  const logger = new ChalkLogger(
    {
      name: 'demo7',
      level: Level.DEBUG,
      flags: {
        date: true,
        title: false,
        colorful: true,
        inline: true,
      },
    },
    process.argv,
  )

  logger.debug('A', 'B', 'C')
  logger.verbose('A', 'B', 'C')
  logger.info('a', 'b', 'c')
  logger.warn('X', 'Y', 'Z', { a: 1, b: 2 })
  logger.error('x', 'y', 'z', { c: { a: 'hello' }, b: { d: 'world' } })
  logger.fatal('1', '2', '3')

  const logger2 = new ChalkLogger(
    {
      name: 'demo7',
      level: Level.DEBUG,
      flags: {
        date: false,
        title: false,
        colorful: true,
        inline: true,
      },
      placeholderRegex: /(?<!\\)<>/g, // change placeholder of string format
    },
    process.argv,
  )

  logger2.debug('A', 'B', 'C')
  logger2.verbose('A', 'B', 'C')
  logger2.info('a', 'b', 'c')
  logger2.warn('X', 'Y', 'Z', { a: 1, b: 2 })
  logger2.error('x', 'y', 'z', { c: { a: 'hello' }, b: { d: 'world' } })
  logger2.fatal('1', '2', '3')
  ```

  ![demo7.1.png][]


[homepage]: https://github.com/guanghechen/node-scaffolds/tree/release-4.x.x/packages/chalk-logger#readme
[@yozora/helper-jest]: https://www.npmjs.com/package/@guanghechen/helper-jest
[demo1.1.png]: https://raw.githubusercontent.com/guanghechen/node-scaffolds/main/packages/chalk-logger/screenshots/demo1.1.png
[demo2.1.png]: https://raw.githubusercontent.com/guanghechen/node-scaffolds/main/packages/chalk-logger/screenshots/demo2.1.png
[demo3.1.png]: https://raw.githubusercontent.com/guanghechen/node-scaffolds/main/packages/chalk-logger/screenshots/demo3.1.png
[demo4.1.png]: https://raw.githubusercontent.com/guanghechen/node-scaffolds/main/packages/chalk-logger/screenshots/demo4.1.png
[demo5.1.png]: https://raw.githubusercontent.com/guanghechen/node-scaffolds/main/packages/chalk-logger/screenshots/demo5.1.png
[demo6.1.png]: https://raw.githubusercontent.com/guanghechen/node-scaffolds/main/packages/chalk-logger/screenshots/demo6.1.png
[demo7.1.png]: https://raw.githubusercontent.com/guanghechen/node-scaffolds/main/packages/chalk-logger/screenshots/demo7.1.png
[chalk]: https://github.com/chalk/chalk
[commander.js]: https://github.com/tj/commander.js
