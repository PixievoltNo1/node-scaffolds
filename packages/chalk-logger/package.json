{
  "name": "@guanghechen/chalk-logger",
  "version": "4.0.0-alpha.7",
  "description": "A simple colorful logger",
  "author": {
    "name": "guanghechen",
    "url": "https://github.com/guanghechen/"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/guanghechen/node-scaffolds/tree/release-4.x.x",
    "directory": "packages/chalk-logger"
  },
  "homepage": "https://github.com/guanghechen/node-scaffolds/tree/release-4.x.x/packages/chalk-logger#readme",
  "keywords": [
    "logger",
    "colorful logger"
  ],
  "type": "module",
  "module": "./lib/esm/index.js",
  "exports": {
    "import": "./lib/esm/index.js"
  },
  "types": "lib/types/index.d.ts",
  "source": "src/index.ts",
  "license": "MIT",
  "engines": {
    "node": ">= 16.0.0"
  },
  "files": [
    "lib/",
    "src/",
    "package.json",
    "CHANGELOG.md",
    "LICENSE",
    "README.md"
  ],
  "scripts": {
    "build": "rimraf lib/ && cross-env NODE_ENV=production rollup -c ../../rollup.config.mjs",
    "prepublishOnly": "yarn build",
    "test": "cross-env TS_NODE_FILES=true NODE_OPTIONS=--experimental-vm-modules jest --config ../../jest.config.mjs --rootDir .",
    "ts:run": "cross-env NODE_ENV=production node --no-warnings --experimental-specifier-resolution=node --loader ts-node/esm -r tsconfig-paths/register"
  },
  "dependencies": {
    "@guanghechen/helper-is": "^4.0.0-alpha.7",
    "@guanghechen/invariant": "^4.0.0-alpha.7",
    "@guanghechen/utility-types": "^4.0.0-alpha.7",
    "chalk": "^5.2.0",
    "dayjs": "^1.11.7",
    "json5": "^2.2.3"
  },
  "devDependencies": {
    "cross-env": "^7.0.3",
    "is-ci": "^3.0.1",
    "jest": "^29.4.2",
    "rimraf": "^4.1.2",
    "rollup": "^3.14.0"
  }
}
