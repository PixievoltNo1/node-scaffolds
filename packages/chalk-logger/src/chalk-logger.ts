import type { Mutable } from '@guanghechen/types'
import { parseOptionsFromArgs } from './commander'
import type { ILoggerOptions } from './logger'
import { Logger } from './logger'

export class ChalkLogger extends Logger {
  /**
   * prefix of logger.name
   */
  protected basename: string | null = null
  /**
   * name passed into .setName()
   */
  protected divisionName: string | null = null

  constructor(options?: ILoggerOptions, args?: string[]) {
    const optionsFromArgs = parseOptionsFromArgs(args || [])
    const resolvedOptions = {
      ...options,
      ...optionsFromArgs,
      flights: { ...options?.flights, ...optionsFromArgs.flights },
    }

    super(resolvedOptions)
    const basename = resolvedOptions?.name ?? ''
    this.setBaseName(basename)
  }

  /**
   * update logger's name
   * @param name
   */
  public setName(name: string | null): void {
    const resolvedName: string = [this.basename, name]
      .filter((x): x is string => x != null && x.length > 0)
      .join(' ')
    const self = this as Mutable<this>
    self.name = resolvedName
    this.divisionName = name
  }

  /**
   * update basename of logger
   * @param basename
   */
  public setBaseName(basename: string | null): void {
    this.basename = basename
    this.setName(this.divisionName)
  }

  /**
   * update logger's mode
   * @param mode
   */
  public setMode(mode: 'normal' | 'loose'): void {
    const self = this as Mutable<this>
    self.mode = mode
  }
}
