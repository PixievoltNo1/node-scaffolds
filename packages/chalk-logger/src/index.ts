import type { Mutable } from '@guanghechen/utility-types'
import { partOptionsFromArgs } from './commander'
import type { Level } from './level'
import type { ILoggerOptions } from './logger'
import { Logger } from './logger'

export * from './color'
export * from './commander'
export * from './level'
export * from './logger'

export class ChalkLogger extends Logger {
  /**
   * prefix of logger.name
   */
  protected basename: string | null = null
  /**
   * name passed into .setName()
   */
  protected divisionName: string | null = null

  constructor(options?: ILoggerOptions, args?: string[]) {
    const optionsFromArgs = partOptionsFromArgs(args || [])
    const resolvedOptions = {
      ...options,
      ...optionsFromArgs,
      flags: { ...options?.flags, ...optionsFromArgs.flags },
    }

    super(resolvedOptions)
    const basename = resolvedOptions?.name ?? ''
    this.setBaseName(basename)
  }

  /**
   * update logger's level
   * @param level
   */
  public setLevel(level: Level | null | undefined): void {
    if (level == null) return
    const self = this as Mutable<this>
    self.level = level
  }

  /**
   * update logger's name
   * @param name
   */
  public setName(name: string | null): void {
    const resolvedName: string = [this.basename, name]
      .filter((x): x is string => x != null && x.length > 0)
      .join(' ')
    const self = this as Mutable<this>
    self.name = resolvedName
    this.divisionName = name
  }

  /**
   * update basename of logger
   * @param basename
   */
  public setBaseName(basename: string | null): void {
    this.basename = basename
    this.setName(this.divisionName)
  }

  /**
   * update logger's mode
   * @param mode
   */
  public setMode(mode: 'normal' | 'loose'): void {
    const self = this as Mutable<this>
    self.mode = mode
  }
}

export default ChalkLogger
