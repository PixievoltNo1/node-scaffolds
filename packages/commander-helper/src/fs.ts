import { isFileSync } from '@guanghechen/file-helper'
import fs from 'fs-extra'
import yaml from 'js-yaml'
import path from 'path'

/**
 * Load configuration file with format .json / .yml / .yaml
 *
 * @param filepath
 * @param logger
 */
export async function loadJsonOrYaml(
  filepath: string,
  encoding = 'utf-8',
): Promise<unknown | never> {
  if (!isFileSync(filepath))
    throw new Error(`${filepath} is an invalid file path`)

  const loadContent = (): Promise<string> =>
    fs.readFile(filepath, encoding as BufferEncoding)

  let result: unknown
  const extname = path.extname(filepath)
  switch (extname) {
    case '.json':
      {
        const content: string = await loadContent()
        result = JSON.parse(content)
      }
      break
    case '.yml':
    case '.yaml':
      {
        const content: string = await loadContent()
        result = yaml.load(content, { filename: filepath, json: true })
      }
      break
    default:
      throw new Error(
        `Only files in .json / .yml / .ymal format are supported. filepath(${filepath}`,
      )
  }
  return result
}

/**
 * Load configuration file with format .json / .yml / .yaml  (synchronizing)
 *
 * @param filepath
 * @param logger
 */
export function loadJsonOrYamlSync(
  filepath: string,
  encoding = 'utf-8',
): unknown | never {
  if (!isFileSync(filepath))
    throw new Error(`${filepath} is an invalid file path`)

  const loadContent = (): string =>
    fs.readFileSync(filepath, encoding as BufferEncoding)

  let result: unknown
  const extname = path.extname(filepath)
  switch (extname) {
    case '.json':
      {
        const content: string = loadContent()
        result = JSON.parse(content)
      }
      break
    case '.yml':
    case '.yaml':
      {
        const content: string = loadContent()
        result = yaml.load(content, { filename: filepath, json: true })
      }
      break
    default:
      throw new Error(
        `Only files in .json / .yml / .ymal format are supported. filepath(${filepath}`,
      )
  }
  return result
}
