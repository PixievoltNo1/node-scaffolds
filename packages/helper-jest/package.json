{
  "name": "@guanghechen/helper-jest",
  "version": "4.7.3",
  "description": "A collection of utility functions for jest tests",
  "author": {
    "name": "guanghechen",
    "url": "https://github.com/guanghechen/"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/guanghechen/node-scaffolds/tree/release-4.x.x",
    "directory": "packages/helper-jest"
  },
  "homepage": "https://github.com/guanghechen/node-scaffolds/tree/release-4.x.x/packages/helper-jest#readme",
  "keywords": [],
  "type": "module",
  "main": "./lib/cjs/index.cjs",
  "module": "./lib/esm/index.mjs",
  "exports": {
    "import": "./lib/esm/index.mjs",
    "require": "./lib/cjs/index.cjs"
  },
  "types": "lib/types/index.d.ts",
  "source": "src/index.ts",
  "license": "MIT",
  "engines": {
    "node": ">= 16.0.0"
  },
  "files": [
    "lib/",
    "src/",
    "package.json",
    "CHANGELOG.md",
    "LICENSE",
    "README.md"
  ],
  "scripts": {
    "build": "rimraf lib/ && cross-env NODE_ENV=production rollup -c ../../rollup.config.mjs",
    "prepublishOnly": "yarn build",
    "test": "cross-env TS_NODE_FILES=true NODE_OPTIONS=--experimental-vm-modules jest --config ../../jest.config.mjs --rootDir ."
  },
  "dependencies": {
    "@guanghechen/helper-func": "^4.7.3",
    "@guanghechen/helper-is": "^4.7.3",
    "@jest/globals": "^29.5.0",
    "jest": "^29.5.0"
  },
  "devDependencies": {
    "cross-env": "^7.0.3",
    "rimraf": "^4.4.1",
    "rollup": "^3.20.2"
  }
}
