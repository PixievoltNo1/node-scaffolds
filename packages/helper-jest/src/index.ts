export * from './desensitizer/json'
export * from './desensitizer/types'
export * from './desensitizer/string'
export * from './mock/console'
export * from './mock/logger'
export * from './snapshot/file'
