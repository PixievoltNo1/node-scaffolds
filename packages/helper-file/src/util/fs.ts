import fs from 'node:fs'
import path from 'node:path'

/**
 * Collect all files under the given directory.
 *
 * @param dir
 * @param predicate
 */
export async function collectAllFiles(
  dir: string,
  predicate?: ((p: string, stat: fs.Stats) => Promise<boolean> | boolean) | null,
): Promise<string[]> {
  const results: string[] = []
  await collect(dir)
  return results

  async function collect(filepath: string): Promise<void> {
    const stat = fs.statSync(filepath)
    if (stat.isDirectory()) {
      const filenames = fs.readdirSync(filepath)
      for (const filename of filenames) {
        const nextFilepath = path.join(filepath, filename)
        await collect(nextFilepath)
      }
      return
    }

    if (stat.isFile()) {
      if (predicate == null || (await predicate(filepath, stat))) {
        results.push(filepath)
      }
    }
  }
}

/**
 * Collect all files under the given directory.   (synchronizing)
 *
 * @param dir
 * @param predicate
 */
export function collectAllFilesSync(
  dir: string,
  predicate?: ((p: string, stat: fs.Stats) => boolean) | null,
): string[] {
  const results: string[] = []
  collect(dir)
  return results

  function collect(filepath: string): void {
    const stat = fs.statSync(filepath)
    if (stat.isDirectory()) {
      const filenames = fs.readdirSync(filepath)
      for (const filename of filenames) {
        const nextFilepath = path.join(filepath, filename)
        collect(nextFilepath)
      }
      return
    }
    if (stat.isFile()) {
      if (predicate == null || predicate(filepath, stat)) {
        results.push(filepath)
      }
    }
  }
}
