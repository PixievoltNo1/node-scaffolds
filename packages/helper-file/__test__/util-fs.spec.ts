import { ChalkLogger, Level } from '@guanghechen/chalk-logger'
import { createLoggerMock } from '@guanghechen/helper-jest'
import fs from 'fs-extra'
import { desensitize, locateFixtures } from 'jest.helper'
import {
  collectAllFiles,
  collectAllFilesSync,
  ensureCriticalFilepathExistsSync,
  isDirectorySync,
  isFileSync,
  isNonExistentOrEmpty,
  mkdirsIfNotExists,
} from '../src'

describe('isFileSync', function () {
  test('truthy', function () {
    expect(isFileSync(locateFixtures('basic/config.yml'))).toBe(true)
  })

  test('falsy', function () {
    expect(isFileSync(locateFixtures('basic/config.yml-non-exist'))).toBe(false)
    expect(isFileSync(locateFixtures('basic'))).toBe(false)
    expect(isFileSync(null)).toBe(false)
    expect(isFileSync(undefined as any)).toBe(false)
  })
})

describe('isDirectorySync', function () {
  test('truthy', function () {
    expect(isDirectorySync(locateFixtures('basic'))).toBe(true)
  })

  test('falsy', function () {
    expect(isDirectorySync(locateFixtures('basic/config.yml'))).toBe(false)
    expect(isDirectorySync(locateFixtures('basic-non-exist'))).toBe(false)
    expect(isDirectorySync(null)).toBe(false)
    expect(isDirectorySync(undefined as any)).toBe(false)
  })
})

describe('isNonExistentOrEmpty', function () {
  test('truthy', function () {
    expect(isNonExistentOrEmpty(locateFixtures('basic-non-exist'))).toBe(true)
    expect(isNonExistentOrEmpty(locateFixtures('basic/config.yml-non-exist'))).toBe(true)
  })

  test('falsy', function () {
    expect(isNonExistentOrEmpty(locateFixtures('basic'))).toBe(false)
    expect(isNonExistentOrEmpty(locateFixtures('basic/config.yml'))).toBe(false)
    expect(isNonExistentOrEmpty(null)).toBe(false)
    expect(isNonExistentOrEmpty(undefined as any)).toBe(false)
  })
})

describe('mkdirsIfNotExists', function () {
  test('directory existed', function () {
    const dirpath = locateFixtures('basic')
    expect(fs.existsSync(dirpath)).toBe(true)
    mkdirsIfNotExists(dirpath, true)
    expect(fs.existsSync(dirpath)).toBe(true)
  })

  test('filepath', function () {
    const dirpath = locateFixtures('basic/config.yml')
    expect(fs.existsSync(dirpath)).toBe(true)
    mkdirsIfNotExists(dirpath, false)
    expect(fs.existsSync(dirpath)).toBe(true)
  })

  test('mkdirs', function () {
    const dirpath = locateFixtures('basic--non-existed')
    expect(fs.existsSync(dirpath)).toBe(false)
    mkdirsIfNotExists(dirpath, true)
    expect(fs.existsSync(dirpath)).toBe(true)
    fs.rmdirSync(dirpath)
  })

  test('mkdirs logger', function () {
    const logger = new ChalkLogger({
      level: Level.VERBOSE,
      flags: {
        colorful: false,
      },
    })
    const loggerMock = createLoggerMock({ logger, desensitize })

    const dirpath = locateFixtures('basic--non-existed--2')
    expect(fs.existsSync(dirpath)).toBe(false)
    mkdirsIfNotExists(dirpath, true, logger)
    expect(fs.existsSync(dirpath)).toBe(true)
    fs.rmdirSync(dirpath)

    expect(loggerMock.getIndiscriminateAll()).toMatchSnapshot()
    loggerMock.restore()
  })
})

describe('ensureCriticalFilepathExistsSync', function () {
  test('null / undefined', function () {
    expect(() => void ensureCriticalFilepathExistsSync(null)).toThrow(
      'Invariant failed: Invalid path: null.',
    )
  })

  test('not found', function () {
    expect(
      () =>
        void ensureCriticalFilepathExistsSync(locateFixtures('basic/config.json-non-existed---22')),
    ).toThrow('Invariant failed: Not found:')
  })

  test('not a file', function () {
    expect(() => void ensureCriticalFilepathExistsSync(locateFixtures('basic/'))).toThrow(
      'Invariant failed: Not a file:',
    )
  })

  test('valid', function () {
    expect(
      () => void ensureCriticalFilepathExistsSync(locateFixtures('basic/config.yml')),
    ).not.toThrow()
  })
})

describe('collectAllFiles', function () {
  test('default predicate', async function () {
    expect(desensitize(await collectAllFiles(locateFixtures('basic')))).toMatchSnapshot()
  })

  test('yaml file only', async function () {
    expect(
      desensitize(await collectAllFiles(locateFixtures('basic'), p => /\.(?:yml|yaml)$/.test(p))),
    ).toMatchSnapshot()

    expect(
      desensitize(
        await collectAllFiles(locateFixtures('basic'), p =>
          Promise.resolve(/\.(?:yml|yaml)$/.test(p)),
        ),
      ),
    ).toMatchSnapshot()
  })

  test('collect start from file', async function () {
    expect(desensitize(await collectAllFiles(locateFixtures('basic/config.yml')))).toMatchSnapshot()
  })
})

describe('collectAllFilesSync', function () {
  test('default predicate', function () {
    expect(desensitize(collectAllFilesSync(locateFixtures('basic')))).toMatchSnapshot()
  })

  test('yaml file only', function () {
    expect(
      desensitize(collectAllFilesSync(locateFixtures('basic'), p => /\.(?:yml|yaml)$/.test(p))),
    ).toMatchSnapshot()
  })

  test('collect start from file', function () {
    expect(desensitize(collectAllFilesSync(locateFixtures('basic/config.yml')))).toMatchSnapshot()
  })
})
