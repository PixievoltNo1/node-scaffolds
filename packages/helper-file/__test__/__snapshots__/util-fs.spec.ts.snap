// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectAllFiles collect start from file 1`] = `
Array [
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/basic/config.yml",
]
`;

exports[`collectAllFiles default predicate 1`] = `
Array [
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/basic/1.md",
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/basic/2.md",
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/basic/config.yml",
]
`;

exports[`collectAllFiles yaml file only 1`] = `
Array [
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/basic/config.yml",
]
`;

exports[`collectAllFiles yaml file only 2`] = `
Array [
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/basic/config.yml",
]
`;

exports[`collectAllFilesSync collect start from file 1`] = `
Array [
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/basic/config.yml",
]
`;

exports[`collectAllFilesSync default predicate 1`] = `
Array [
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/basic/1.md",
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/basic/2.md",
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/basic/config.yml",
]
`;

exports[`collectAllFilesSync yaml file only 1`] = `
Array [
  "<$WORKSPACE$>/packages/helper-file/__test__/fixtures/basic/config.yml",
]
`;

exports[`mkdirsIfNotExists mkdirs logger 1`] = `
Array [
  Array [
    "[verb ] mkdirs: <$WORKSPACE$>/packages/helper-file/__test__/fixtures/basic--non-existed--2
",
  ],
]
`;
